<h1> Data Exercise #4 | Data Collection on Used Cars</h1>
<p><b>4.</b> Collect data on used cars of a specific make and model (e.g., Ford Focus) in a city, from the Web,
using a classified ads website or a used cars website as a source. Use web scraping to collect all
available data on these cars. (Alternatively, collect the most important variables by hand from the
100 most recently advertised cars.) Write a short report on what you did, how many cars you
ended up with in the data, and what difficulties you encountered, if any. </p>
<h3>Answer:</h3>
<p>This was a very challenging exercise, technically speaking. I went though several tutorials and documentations to get the desired data. I'll describe part of the process together with a brief analytical content of it.</p>
<p>My goal was to get all ads for <b>Hyundai</b> <b>HB20</b> models in the state of <b>Rio Grande do Sul</b>, <b>Brazil</b>. The extraction used in this notebook was performed in <b>March 7h, 2023</b>. I consulted the <a href='https://www.olx.com.br/'>OLX</a> online ad website to get the data (thank you OLX!).</p>
<p>If you'd like to know how I scraped the data, please refer to the <code>.py</code> app stored in this same folder. I'll be commenting some of the difficulties that I'd faced when trying to develop this program.</p>
<p>First, I'll import the required libraries and load the dataset.</p>

In [1]:
import os
import pandas as pd

In [7]:
current_path = os.getcwd()
data_in = f"{current_path}/data/output/"

In [4]:
df_cars = pd.read_csv(f"{data_in}WS_car_prices.csv")

In [8]:
df_cars.shape

(704, 11)

<p>As we can see, I managed to scrape 704 ads, all of them containing 11 variables that qualify each observed data. Let's take a look at each variable.</p>

In [9]:
df_cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 704 entries, 0 to 703
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   car_model      704 non-null    object 
 1   car_price      704 non-null    float64
 2   date           704 non-null    object 
 3   time           704 non-null    object 
 4   total_mileage  704 non-null    float64
 5   year           704 non-null    int64  
 6   gear_type      704 non-null    object 
 7   fuel_type      704 non-null    object 
 8   url            704 non-null    object 
 9   city           704 non-null    object 
 10  area           704 non-null    object 
dtypes: float64(2), int64(1), object(8)
memory usage: 60.6+ KB


<p>As we can see, I scraped a good deal of information on each ad. Although the data must still go through an exploratory and a cleaning step, we can already observe that, besides a title for each ad, one can get the following information:
<ul>
    <li>price</li>
    <li>date of the ad</li>
    <li>time when the ad was updated</li>
    <li>mileage</li>
    <li>url for each ad</li>
    <li>gear and fuel type</li>
    <li>city and city area</li>
</ul>

<h4>Some of the difficulties that I'd faced during the web scraping process:</h4>
The first problem was that, today, most ad websites work on dynamic content. That is, the content for each ad is dynamically updated via a backend js application, which updates the date after the webpage is accessed. Hence, traditionally-employed libraries such as <code>requests</code> only return empty values when accessed via a <code>BeautifulSoup</code> text parser. 
<b>The solution</b> came with a different library: </code>Selenium</code>, which is primarily used to run tests on a web browser. In my case, I used it to simulate a series of visits through each page of the website. For every page that the driver would open, the relevant data would be accessed, parsed, and stored in a <code>json</code>, which would be later converted to a <code>pandas DataFrame</code> so that I could adjust some issues and then export the extracted data to a <code>csv</code> file. Therefore, although I had to solve other problems (which are all documented in the app), the most difficult part was to actually get hold of the data; the rest was a series of smaller issues.</p>
<hr>